File: tests/scalers/aws-kinesis-stream.test.ts
import test from 'ava'
import * as fs from 'fs'
import * as sh from 'shelljs'
import * as tmp from 'tmp'
import {v4 as uuidv4} from 'uuid';
import {
  Kinesis,
} from '@aws-sdk/client-kinesis'
import { createNamespace, sleep, waitForDeploymentReplicaCount } from './helpers'


const awsRegion = 'eu-west-2'
const awsAccessKey = process.env['AWS_ACCESS_KEY'];
const awsSecretKey =  process.env['AWS_SECRET_KEY'];
const testNamespace = 'kinesis-test'
const nginxDeploymentName = 'nginx-deployment'
const streamName = `keda-stream-${uuidv4()}`

let kinesisClient: Kinesis;

const updateShardCountAsync = (count: number) => new Promise((resolve, _) => {
  kinesisClient.updateShardCount({
    StreamName:streamName,
    TargetShardCount: count,
    ScalingType:'UNIFORM_SCALING'
  }, async err => {
    if (err != null) console.log(err)
    // Wait till the stream is updated and ready
    await sleep(30000)
    resolve(undefined);
  })
});

test.before(async t => {
  // setup kinesis client
  kinesisClient = new Kinesis({
    region: awsRegion,
    credentials: {
      accessKeyId: awsAccessKey,
      secretAccessKey: awsSecretKey
    }
  });

  createNamespace(testNamespace)

  // create the stream
  const createStreamAsync = () => new Promise((resolve, _) => {
    var params = {
      'ShardCount': 1,
      'StreamName': streamName
    };
    kinesisClient.createStream(params, async err => {
      if (err != null) console.log(err)
      // Wait till the stream is created and ready
      await sleep(30000)
      resolve(undefined);
    })
  });
  await createStreamAsync()


  // deploy nginx, scaledobject etc.
  console.log('deploy nginx, scaledobject etc.')
  const nginxTmpFile = tmp.fileSync()
  fs.writeFileSync(nginxTmpFile.name, nginxDeployYaml)

  t.is(0, sh.exec(`kubectl apply --namespace ${testNamespace} -f ${nginxTmpFile.name}`).code, 'creating nginx deployment should work.')
})

test.serial(`Replicas should scale to 2 (the max) then back to 0`, async t => {
  const maxReplicaCount = 2
  const minReplicaCount = 1;

  t.true(await waitForDeploymentReplicaCount(minReplicaCount, nginxDeploymentName, testNamespace, 300, 1000), 'Replica count should start out as 1')

  await updateShardCountAsync(2)

  t.true(await waitForDeploymentReplicaCount(maxReplicaCount, nginxDeploymentName, testNamespace, 180, 1000), 'Replica count should increase to the maxReplicaCount')

  await updateShardCountAsync(1)

  t.true(await waitForDeploymentReplicaCount(minReplicaCount, nginxDeploymentName, testNamespace, 180, 1000), 'Replica count should increase to the minReplicaCount')

})

test.after.always(async (t) => {
  // delete the stream
  const deleteStreamAsync = () => new Promise((resolve, _) => {
    var params = {
      'StreamName': streamName
    };
    kinesisClient.deleteStream(params, err => {
      if (err != null) console.log(err)
      resolve(undefined);
    })
  });
  await deleteStreamAsync()

  t.is(0, sh.exec(`kubectl delete namespace ${testNamespace}`).code, 'Should delete kinesis namespace')
})

const nginxDeployYaml = `
apiVersion: v1
kind: Secret
metadata:
  name: test-secrets
data:
  AWS_ACCESS_KEY_ID: '${Buffer.from(awsAccessKey, 'binary').toString('base64')}' # Required.
  AWS_SECRET_ACCESS_KEY: '${Buffer.from(awsSecretKey, 'binary').toString('base64')}' # Required.
---
apiVersion: keda.sh/v1alpha1
kind: TriggerAuthentication
metadata:
  name: keda-trigger-auth-aws-credentials
spec:
  secretTargetRef:
  - parameter: awsAccessKeyID     # Required.
    name: test-secrets            # Required.
    key: AWS_ACCESS_KEY_ID        # Required.
  - parameter: awsSecretAccessKey # Required.
    name: test-secrets            # Required.
    key: AWS_SECRET_ACCESS_KEY    # Required.
---
apiVersion: keda.sh/v1alpha1
kind: ScaledObject
metadata:
  name: aws-kinesis-so
  labels:
    app: nginx
spec:
  scaleTargetRef:
    name: ${nginxDeploymentName}
  maxReplicaCount: 2
  minReplicaCount: 1
  cooldownPeriod: 1
  advanced:
    horizontalPodAutoscalerConfig:
      behavior:
        scaleDown:
          stabilizationWindowSeconds: 15
  triggers:
    - type: aws-kinesis-stream
      authenticationRef:
        name: keda-trigger-auth-aws-credentials
      metadata:
        awsRegion: ${awsRegion}
        streamName: ${streamName}
        shardCount: "1"
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: ${nginxDeploymentName}
  labels:
    app: nginx
spec:
  replicas: 1
  selector:
    matchLabels:
      app: nginx
  template:
    metadata:
      labels:
        app: nginx
    spec:
      containers:
      - name: nginx
        image: nginx:1.14.2
        ports:
        - containerPort: 80
---
`

File: tests/scalers/aws/aws_kinesis_stream/aws_kinesis_stream_test.go
//go:build e2e
// +build e2e

package aws_kinesis_stream_test

import (
	"context"
	"encoding/base64"
	"fmt"
	"os"
	"testing"
	"time"

	"github.com/aws/aws-sdk-go/aws"
	"github.com/aws/aws-sdk-go/aws/credentials"
	"github.com/aws/aws-sdk-go/aws/session"
	"github.com/aws/aws-sdk-go/service/kinesis"
	"github.com/joho/godotenv"
	"github.com/stretchr/testify/assert"
	"k8s.io/client-go/kubernetes"

	. "github.com/kedacore/keda/v2/tests/helper"
)

// Load environment variables from .env file
var _ = godotenv.Load("../../../.env")

const (
	testName = "aws-kinesis-stream-test"
)

type templateData struct {
	TestNamespace      string
	DeploymentName     string
	ScaledObjectName   string
	SecretName         string
	AwsAccessKeyID     string
	AwsSecretAccessKey string
	AwsRegion          string
	KinesisStream      string
}

const (
	secretTemplate = `apiVersion: v1
kind: Secret
metadata:
  name: {{.SecretName}}
  namespace: {{.TestNamespace}}
data:
  AWS_ACCESS_KEY_ID: {{.AwsAccessKeyID}}
  AWS_SECRET_ACCESS_KEY: {{.AwsSecretAccessKey}}
`

	triggerAuthenticationTemplate = `apiVersion: keda.sh/v1alpha1
kind: TriggerAuthentication
metadata:
  name: keda-trigger-auth-aws-credentials
  namespace: {{.TestNamespace}}
spec:
  secretTargetRef:
  - parameter: awsAccessKeyID     # Required.
    name: {{.SecretName}}         # Required.
    key: AWS_ACCESS_KEY_ID        # Required.
  - parameter: awsSecretAccessKey # Required.
    name: {{.SecretName}}         # Required.
    key: AWS_SECRET_ACCESS_KEY    # Required.
`

	deploymentTemplate = `
apiVersion: apps/v1
kind: Deployment
metadata:
  name: {{.DeploymentName}}
  namespace: {{.TestNamespace}}
  labels:
    app: {{.DeploymentName}}
spec:
  replicas: 0
  selector:
    matchLabels:
      app: {{.DeploymentName}}
  template:
    metadata:
      labels:
        app: {{.DeploymentName}}
    spec:
      containers:
      - name: nginx
        image: nginx:1.14.2
        ports:
        - containerPort: 80
`

	scaledObjectTemplate = `
apiVersion: keda.sh/v1alpha1
kind: ScaledObject
metadata:
  name: {{.ScaledObjectName}}
  namespace: {{.TestNamespace}}
  labels:
    app: {{.DeploymentName}}
spec:
  scaleTargetRef:
    name: {{.DeploymentName}}
  maxReplicaCount: 2
  minReplicaCount: 0
  cooldownPeriod: 1
  advanced:
    horizontalPodAutoscalerConfig:
      behavior:
        scaleDown:
          stabilizationWindowSeconds: 15
  triggers:
    - type: aws-kinesis-stream
      authenticationRef:
        name: keda-trigger-auth-aws-credentials
      metadata:
        awsRegion: {{.AwsRegion}}
        streamName: {{.KinesisStream}}
        shardCount: "3"
        activationShardCount: "4"
`
)

var (
	testNamespace      = fmt.Sprintf("%s-ns", testName)
	deploymentName     = fmt.Sprintf("%s-deployment", testName)
	scaledObjectName   = fmt.Sprintf("%s-so", testName)
	secretName         = fmt.Sprintf("%s-secret", testName)
	kinesisStreamName  = fmt.Sprintf("kinesis-%d", GetRandomNumber())
	awsAccessKeyID     = os.Getenv("TF_AWS_ACCESS_KEY")
	awsSecretAccessKey = os.Getenv("TF_AWS_SECRET_KEY")
	awsRegion          = os.Getenv("TF_AWS_REGION")
	maxReplicaCount    = 2
	minReplicaCount    = 0
)

func TestKiensisScaler(t *testing.T) {
	// setup kinesis
	kinesisClient := createKinesisClient()
	createKinesisStream(t, kinesisClient)

	// Create kubernetes resources
	kc := GetKubernetesClient(t)
	data, templates := getTemplateData()
	CreateKubernetesResources(t, kc, testNamespace, data, templates)

	assert.True(t, WaitForDeploymentReplicaReadyCount(t, kc, deploymentName, testNamespace, minReplicaCount, 60, 1),
		"replica count should be %d after 1 minute", minReplicaCount)

	// test scaling
	testActivation(t, kc, kinesisClient)
	testScaleOut(t, kc, kinesisClient)
	testScaleIn(t, kc, kinesisClient)

	// cleanup
	DeleteKubernetesResources(t, kc, testNamespace, data, templates)
	cleanupStream(t, kinesisClient)
}

func testActivation(t *testing.T, kc *kubernetes.Clientset, kinesisClient *kinesis.Kinesis) {
	t.Log("--- testing activation ---")
	updateShardCount(t, kinesisClient, 3)
	AssertReplicaCountNotChangeDuringTimePeriod(t, kc, deploymentName, testNamespace, minReplicaCount, 60)
}

func testScaleOut(t *testing.T, kc *kubernetes.Clientset, kinesisClient *kinesis.Kinesis) {
	t.Log("--- testing scale out ---")
	updateShardCount(t, kinesisClient, 6)
	assert.True(t, WaitForDeploymentReplicaReadyCount(t, kc, deploymentName, testNamespace, maxReplicaCount, 60, 3),
		"replica count should be %d after 3 minutes", maxReplicaCount)
}

func testScaleIn(t *testing.T, kc *kubernetes.Clientset, kinesisClient *kinesis.Kinesis) {
	t.Log("--- testing scale in ---")
	updateShardCount(t, kinesisClient, 3)

	assert.True(t, WaitForDeploymentReplicaReadyCount(t, kc, deploymentName, testNamespace, minReplicaCount, 60, 3),
		"replica count should be %d after 3 minutes", minReplicaCount)
}

func updateShardCount(t *testing.T, kinesisClient *kinesis.Kinesis, shardCount int64) {
	done := waitForStreamActiveStatus(t, kinesisClient)
	if done {
		_, err := kinesisClient.UpdateShardCountWithContext(context.Background(), &kinesis.UpdateShardCountInput{
			StreamName:       &kinesisStreamName,
			TargetShardCount: aws.Int64(shardCount),
			ScalingType:      aws.String("UNIFORM_SCALING"),
		})
		assert.NoErrorf(t, err, "cannot update shard count - %s", err)
	}
	assert.True(t, true, "failed to update shard count")
}

func createKinesisStream(t *testing.T, kinesisClient *kinesis.Kinesis) {
	_, err := kinesisClient.CreateStreamWithContext(context.Background(), &kinesis.CreateStreamInput{
		StreamName: &kinesisStreamName,
		ShardCount: aws.Int64(2),
	})
	assert.NoErrorf(t, err, "failed to create stream - %s", err)
	done := waitForStreamActiveStatus(t, kinesisClient)
	if !done {
		assert.True(t, true, "failed to create kinesis")
	}
}

func waitForStreamActiveStatus(t *testing.T, kinesisClient *kinesis.Kinesis) bool {
	for i := 0; i < 30; i++ {
		describe, _ := kinesisClient.DescribeStreamWithContext(context.Background(), &kinesis.DescribeStreamInput{
			StreamName: &kinesisStreamName,
		})
		t.Logf("Waiting for stream ACTIVE status. current status - %s", *describe.StreamDescription.StreamStatus)
		if *describe.StreamDescription.StreamStatus == "ACTIVE" {
			return true
		}
		time.Sleep(time.Second * 2)
	}
	return false
}

func cleanupStream(t *testing.T, kinesisClient *kinesis.Kinesis) {
	t.Log("--- cleaning up ---")
	_, err := kinesisClient.DeleteStreamWithContext(context.Background(), &kinesis.DeleteStreamInput{
		StreamName: &kinesisStreamName,
	})
	assert.NoErrorf(t, err, "cannot delete stream - %s", err)
}

func createKinesisClient() *kinesis.Kinesis {
	sess := session.Must(session.NewSession(&aws.Config{
		Region: aws.String(awsRegion),
	}))

	return kinesis.New(sess, &aws.Config{
		Region:      aws.String(awsRegion),
		Credentials: credentials.NewStaticCredentials(awsAccessKeyID, awsSecretAccessKey, ""),
	})
}

func getTemplateData() (templateData, []Template) {
	return templateData{
			TestNamespace:      testNamespace,
			DeploymentName:     deploymentName,
			ScaledObjectName:   scaledObjectName,
			SecretName:         secretName,
			AwsAccessKeyID:     base64.StdEncoding.EncodeToString([]byte(awsAccessKeyID)),
			AwsSecretAccessKey: base64.StdEncoding.EncodeToString([]byte(awsSecretAccessKey)),
			AwsRegion:          awsRegion,
			KinesisStream:      kinesisStreamName,
		}, []Template{
			{Name: "secretTemplate", Config: secretTemplate},
			{Name: "triggerAuthenticationTemplate", Config: triggerAuthenticationTemplate},
			{Name: "deploymentTemplate", Config: deploymentTemplate},
			{Name: "scaledObjectTemplate", Config: scaledObjectTemplate},
		}
}
